### PR TITLE
Cache FreeType faces instead of loading new

### PR DESCRIPTION
**Pull request by kchibisov**
This commit introduces a cache for all loading faces, and replaces `Face` with `FaceLoadingProperties`, since we don't need to carry faces around, only loading options for them. It also removes `FontID` in favor of building `FontKey` right away.

I'd also like to test an idea of returning `FaceLoadingProperties` over `FontKey`, however it'll likely require to use `Rc` boxes or cloning them. So I'd like to test performance of the solution mentioned in this paragraph.

In addition, I think it'll be nice to add an API for rasterizer to remove curtain keys from use, since now they're kind of persistent, but I'm not sure that it's up to this PR.

### CODE REVIEW

File: font/src/ft/mod.rs [hunk 1]
```diff
@@ -63,22 +65,20 @@ struct FallbackList {
...
-    ft_face: freetype::Face,
-    key: FontKey,
+struct FaceLoadingProperties {
     load_flags: freetype::face::LoadFlag,
     render_mode: freetype::RenderMode,
     lcd_filter: c_uint,
     non_scalable: Option<FixedSize>,
     has_color: bool,
     pixelsize_fixup_factor: Option<f64>,
+    ft_face: Rc<FTFace>,
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]: Okay, before going deeper into this code I have the question what the benefit is of having a reference to the face in the hashmap here, rather than just storing the PathBuf to the font?

Is the goal solely to make it so the reference is dropped, as soon as `FaceLoadingProperties` is dropped, destroying the face?

Are there any concerns about just trashing the `ft_faces` hashmap with a bazillion paths to faces that don't even exist anymore?

Comment (kchibisov) [reply font/src/ft/mod.rs]: >Is the goal solely to make it so the reference is dropped, as soon as FaceLoadingProperties is dropped, destroying the face

Yes, that was the idea.

>Is the goal solely to make it so the reference is dropped, as soon as FaceLoadingProperties is dropped, destroying the face

IIRC paths are cheaper to keep around compared to faces, since it’s max 4KB per path, but it is a worst case.

Comment (chrisduerr) [reply font/src/ft/mod.rs]: Have you ever checked how much is needed to stare a face in Rust? 4KB seems like a lot.

Comment (kchibisov) [reply font/src/ft/mod.rs]: Oh, in addition having reference to a `FTFace` simplify the logic, since you need `FTFace` to perform rasterizing, so using indirection with `PathBuf` is slower, since you need to query `HashMap` each time to get a proper face. I have an idea on how to solve your zombie `PathBuf` issue though.

Comment (kchibisov) [reply font/src/ft/mod.rs]: Ok I was a bit wrong that I know a way to remove this zombies automatically. However you repro an issue I'm fixing by doing `cat /dev/urandom` and starting changing your font size, you should notice that master uses way more RAM on average.

File: font/src/ft/mod.rs [hunk 2]
```diff
@@ -32,28 +34,28 @@ use super::{
     Style, Weight,
 };
 
+#[derive(Debug, PartialEq, Copy, Clone)]
 struct FixedSize {
     pixelsize: f64,
 }
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]: Unrelated to this PR directly, but since there's a change to this: Why is this a dedicated struct? Seems like we could just as well use a `f64` directly?

File: font/src/ft/mod.rs [hunk 3]
```diff
@@ -253,31 +253,23 @@ impl FreeTypeRasterizer {
...
+                    self.faces.remove(&fallback_font.key);
                 }
             }
-
-            let _ = self.faces.remove(&font_key);
-            Some(font_key)
-        } else {
-            None
-        };
+        }
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]: The `primary_font_key` is a hash that uniquely identifies a font, since it's just a combinations of all patterns required to load it.

So it seems like we should just be able to return the `FontKey` immediately when we find that it is already loaded? Or what do we reload the fonts for?

File: font/src/ft/mod.rs [hunk 4]
```diff
@@ -302,32 +294,55 @@ impl FreeTypeRasterizer {
...
-
-        let width = match face.ft_face.load_char('0' as usize, face.load_flags) {
-            Ok(_) => face.ft_face.glyph().metrics().horiAdvance / 64,
-            Err(_) => size_metrics.max_advance / 64,
-        } as f64;
+        let width =
+            match face_load_props.ft_face.load_char('0' as usize, face_load_props.load_flags) {
+                Ok(_) => face_load_props.ft_face.glyph().metrics().horiAdvance / 64,
+                Err(_) => size_metrics.max_advance / 64,
+            } as f64;
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]: Might be worth to extract `face_load_props.ft_face` into a variable here, since it would prevent two very ugly linebreaks.

File: font/src/ft/mod.rs [hunk 5]
```diff
@@ -302,32 +294,55 @@ impl FreeTypeRasterizer {
...
+
+            let ft_face = if let Some(face) = self.ft_faces.get(&path) {
+                if let Some(ft_face) = face.upgrade() {
+                    ft_face
+                } else {
+                    self.load_ft_face(path, index)?
+                }
+            } else {
+                self.load_ft_face(path, index)?
+            };
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]: Maybe we can do something like this here:

```rust
let face = match self.ft_faces.get(&path).and_then(|face| face.upgrade()).unwrap_or_else(|| self.load_ft_face(path, index)?;
```

I'm not sure about the exact syntax, but it should work. Alternatively if it looks nicer, we could also `match` inside of `unwrap_or_else`, but I think it should format nicely.

Comment (kchibisov) [reply font/src/ft/mod.rs]: It doesn't look nice, since you can't use `?` in a closure, you also can't use `ok_or_else`, since it applies random thing and you don't have a proper function to do something if none, but not an `unwrap`, like `map_none` could help, but we don't have one, so you should add it youself, which isn't helpful here and you can't do it un Rust anyways IIRC.

Comment (chrisduerr) [reply font/src/ft/mod.rs]: Could still apply the `match` suggestion just fine.

Comment (kchibisov) [reply font/src/ft/mod.rs]: Oh, it was really hard to parse your one liner tbh.

File: font/src/ft/mod.rs [hunk 6]
```diff
@@ -306,32 +285,49 @@ impl FreeTypeRasterizer {
...
-            Ok(_) => face.ft_face.glyph().metrics().horiAdvance / 64,
+        let width = match ft_face.load_char('0' as usize, face_load_props.load_flags) {
+            Ok(_) => ft_face.glyph().metrics().horiAdvance / 64,
             Err(_) => size_metrics.max_advance / 64,
         } as f64;
 
         Ok(FullMetrics { size_metrics, cell_width: width })
     }
 
+    fn load_ft_face(&mut self, path: PathBuf, index: isize) -> Result<Rc<FTFace>, Error> {
```
Comment (kchibisov) [inline font/src/ft/mod.rs]: We can inline this function now, what do you think about it?

Comment (chrisduerr) [reply font/src/ft/mod.rs]: I don't see anything wrong with keeping it separate.

File: font/src/ft/mod.rs [hunk 7]
```diff
@@ -253,31 +246,16 @@ impl FreeTypeRasterizer {
...
 
-            let _ = self.faces.remove(&font_key);
-            Some(font_key)
-        } else {
-            None
-        };
+        // Return font if we already loaded it
+        if self.faces.get(&primary_font_key).is_some() {
+            return Ok(primary_font_key);
+        }
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]: Thinking about this early return, there might be an issue with fallback faces, right?

In `self.faces` both primary and fallback faces are stored. However, a primary face requires the additional computation we do in this method to load the fallback list.

So it seems like what we should do is check if the font is already loaded **and** if a fallback list exists for this font.

However even if the fallback list does not exist, we should be able to still skip the call to `face_from_pattern`. Just not the entire rest of this function.

Comment (kchibisov) [reply font/src/ft/mod.rs]: >Thinking about this early return, there might be an issue with fallback faces, right?

Yeah, exactly, forgot about this case.

File: font/src/ft/mod.rs [hunk 8]
```diff
@@ -253,82 +246,91 @@ impl FreeTypeRasterizer {
...
-                        self.keys.remove(&fallback_font.id);
-                    }
-                }
-            }
+        let primary_font_key = FontKey::from_pattern_hashes(hash, primary_font.hash());
 
-            let _ = self.faces.remove(&font_key);
-            Some(font_key)
+        // Return font if we already loaded it
+        if self.faces.get(&primary_font_key).is_some() {
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]:
```suggestion
        if self.fallback_lists.get(&primary_font_key).is_some() {
```

File: font/src/ft/mod.rs [hunk 9]
```diff
@@ -253,82 +246,91 @@ impl FreeTypeRasterizer {
...
+            self.face_from_pattern(&primary_font, primary_font_key)
+                .and_then(|pattern| pattern.ok_or_else(|| Error::MissingFont(desc.to_owned())))?
         };
 
-        // Reuse the font_key, since changing it can break library users
-        let font_key = self
-            .face_from_pattern(&primary_font, primary_font_id, font_key)
-            .and_then(|pattern| pattern.ok_or_else(|| Error::MissingFont(desc.to_owned())))?;
+        // Build fallback list if we don't have one for a given font key
+        if self.fallback_lists.get(&primary_font_key).is_none() {
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]: Don't need to check that here with my earlier suggestion.

File: font/src/ft/mod.rs [hunk 10]
```diff
@@ -248,21 +248,16 @@ impl FreeTypeRasterizer {
         // Hash pattern together with request pattern to include requested font size in the hash
         let primary_font_key = FontKey::from_pattern_hashes(hash, primary_font.hash());
 
-        // Return font if we already loaded it
-        if self.faces.get(&primary_font_key).is_some() {
-            return Ok(primary_font_key);
-        }
+        // Load font if we haven't loaded one
+        if self.fallback_lists.get(&primary_font_key).is_none() {
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]: Early return here, please `contains_key` instead of actually accessing the value and then calling `is_some()` or `is_none()`.

Comment (kchibisov) [reply font/src/ft/mod.rs]: I did this thing on purpose, because clippy has a bug when using `contains_key` here and then `insert`, like it asks you to use `entry`, but it doesn't make any sense, since you should compute the thing before insert.

File: font/src/ft/mod.rs [hunk 11]
```diff
@@ -253,123 +246,107 @@ impl FreeTypeRasterizer {
...
+        // Return if we already have the same primary font
+        if self.fallback_lists.contains_key(&primary_font_key) {
+            return Ok(primary_font_key);
+        }
 
-        // Reuse the font_key, since changing it can break library users
-        let font_key = self
-            .face_from_pattern(&primary_font, primary_font_id, font_key)
-            .and_then(|pattern| pattern.ok_or_else(|| Error::MissingFont(desc.to_owned())))?;
+        // Load font if we haven't loaded one
```
Comment (chrisduerr) [inline font/src/ft/mod.rs]:
```suggestion
        // Load font if we haven't loaded it yet
```

Since `one` is not ideal here, we only ever want to load one when talking about primary fonts.
